### DISCUSSION TITLE
tasmota script: Array Basic questions

### DISCUSSION THREAD

**Discussion by 2000ede** [arendst/Tasmota#20304] [Q&A]
(update in english)

Hello,
I have questions about arrays in Tasmota Script

I'm struggling through this, but it's difficult (beginner and no English as a native language:
https://tasmota.github.io/docs/Scripting-Language/

1) I actually need arrays with two dimensions
For time, an array with 5 sub-compartments, e.g. days in the month = 31 compartments with 5 sensors each
Multiple arrays then for different times
- last hour in minutes
- hours of the day
- Days of the month
- months of the year

Contents: Counter readings 10 digits including decimal places
Can I make arrays with two dimensions?
Currently I have single level arrays

2) Adopted from predecessor: >D 40
I can't say whether 40 is correct. Does it mean that values ​​in an array can have 40 characters?

3) I would like to first determine the differences (=consumption) in the output based (W) on the counter readings, so that I don't need my own arrays for consumption. (The meter readings are the “truth” and I definitely always want to have them as a basis)
How can I add/subtract variable under W? Probably not possible, right?

4) In Tasmota, can I send all variables including arrays (or an array) via script via json and/or retrieve them with an API command? Currently I only know the API command that sends me the original data from the meter.

5) How can I set the contents of an array to 0 with one command?

6) Currently he only lets me make 4 arrays. If I have more, it won't save (hourglass runs while saving...)
It's not yet clear to me what limitations (memory) etc. I have and how I can proceed more efficiently.

ESP Chip ID 2922968 (ESP32-D0WD-V3 v3.1)
Flash Chip ID 0x164054 (DIO)
Flash Size 4096 KB
Program Flash Size 4096 KB
Program Size 1952 KB
Free Program Space 927 KB
Free Memory 130.6 KB (frag. 14%)
Partition safeboot 832 KB (used 92%)
Partition app0* 2880 KB (used 67%)
Partition fs 320 KB

Thanks for your beginner help! Don't judge me too hard!

**Comment by Jason2866** [comment]
Only English questions here.

**Reply by 2000ede** [reply]
@Jason2866 Thanks for the tip.
I apologize! I changed it.

**Comment by sfromis** [comment]
When you are using ESP32, you might want to consider the Berry language, which is more of a standard programming language with the "usual features", somewhat similar to a simplified Python, without the strictures the scripter feature needed due to this being created to fit a lot of functions into the much lower memory ESP8266. With Berry, memory size is the limit, and if you want to use a lot, you can get an ESP32 board with PSRAM, like sizes of 2/4/8/16 MB. In Berry, a list is like an array, and you can nest lists as member of lists.

Preferences can be called a question of style, but personally, I've found the Berry language pleasant to use for processing data on-device.

**Reply by 2000ede** [reply]
thanks for the hint. I've come a long way and would like to wait for the answers to my questions. 

I hope there are experts and practitioners here who can answer it. Thanks!

**Comment by 2000ede** [comment]
Since I haven't made any progress in this ask yet, I came up with the idea of, for example, instead of 4x one array per counter, just an array where I simply extend the area in the array and put 4 counters into the array
